### PR TITLE
Changes to avoid iterator next() calls.

### PR DESCRIPTION
**Pull request by immutablemaps**
Changes AbstractCollection and ImmutableList to avoid next() calls for
iterators whenever possible.

Also adds some utilities to Collections so clients don't have to use
java util collections methods that use iterative next() calls.

### CODE REVIEW

File: drv/AbstractCollection.drv [hunk 1]
```diff
@@ -160,6 +161,42 @@ public abstract class ABSTRACT_COLLECTION KEY_GENERIC extends AbstractCollection
 			}
 		return retVal;
 	}
+
+	@Override
+	public boolean addAll(final Collection<? extends KEY_GENERIC_CLASS> c) {
+		if (c instanceof COLLECTION){
+			return addAll((COLLECTION)c);
```
Comment (yasserg) [inline drv/AbstractCollection.drv]: space before c

File: drv/AbstractCollection.drv [hunk 2]
```diff
@@ -160,6 +161,42 @@ public abstract class ABSTRACT_COLLECTION KEY_GENERIC extends AbstractCollection
...
+		} else {
+			return super.addAll(c);
+		}
+	}
+
+	@Override
+	public boolean containsAll(final Collection<?> c) {
+		if (c instanceof COLLECTION) {
+			return containsAll((COLLECTION)c);
```
Comment (yasserg) [inline drv/AbstractCollection.drv]: same

File: drv/AbstractCollection.drv [hunk 3]
```diff
@@ -160,6 +161,42 @@ public abstract class ABSTRACT_COLLECTION KEY_GENERIC extends AbstractCollection
...
+		} else {
+			return super.containsAll(c);
+		}
+	}
+
+	@Override
+	public boolean removeAll(final Collection<?> c) {
+		if(c instanceof COLLECTION) {
```
Comment (yasserg) [inline drv/AbstractCollection.drv]: space after if

File: drv/AbstractCollection.drv [hunk 4]
```diff
@@ -160,6 +161,42 @@ public abstract class ABSTRACT_COLLECTION KEY_GENERIC extends AbstractCollection
...
+			return removeAll((COLLECTION)c);
+		} else {
+			return super.removeAll(c);
+		}
+	}
+
+	@Override
+	public boolean retainAll(final Collection<?> c) {
+		if (c instanceof COLLECTION) {
+			return retainAll((COLLECTION)c);
```
Comment (yasserg) [inline drv/AbstractCollection.drv]: same

File: drv/Collections.drv [hunk 5]
```diff
@@ -371,6 +372,40 @@ public final class COLLECTIONS {
 	}
 
 
+#if KEYS_PRIMITIVE && defined JDK_PRIMITIVE_ITERATOR
+	/** Sorts the specified list in ascending order.
```
Comment (yasserg) [inline drv/Collections.drv]: move description to next line

File: drv/Collections.drv [hunk 6]
```diff
@@ -371,6 +372,40 @@ public final class COLLECTIONS {
...
+        Arrays.sort(arr);
+        KEY_LIST_ITERATOR itr = l.listIterator();
+        for (KEY_TYPE e : arr) {
+            itr.NEXT_KEY();
+            itr.set(e);
+        }
+    }
+#endif
+
+#if KEYS_PRIMITIVE && defined JDK_PRIMITIVE_ITERATOR
```
Comment (yasserg) [inline drv/Collections.drv]: merge with above?

File: drv/Collections.drv [hunk 7]
```diff
@@ -371,6 +372,40 @@ public final class COLLECTIONS {
...
+	/** Returns the minimum element in given collection. The given collection must be non-empty.
```
Comment (yasserg) [inline drv/Collections.drv]: same

File: drv/ImmutableList.drv [hunk 8]
```diff
@@ -94,11 +104,27 @@ public abstract class IMMUTABLE_LIST extends ABSTRACT_LIST {
...
             data = new KEY_TYPE[collection.size()];
+#if KEYS_PRIMITIVE
+            if (collection instanceof ABSTRACT_COLLECTION) {
+                KEY_ITERATOR iterator = ((ABSTRACT_COLLECTION)collection).iterator();
+                int pos = 0;
+                while (iterator.hasNext()) {
+                    data[pos++] = iterator.NEXT_KEY();
+                }
+            } else {
+                Iterator<KEY_GENERIC_CLASS> iterator = collection.iterator();
```
Comment (yasserg) [inline drv/ImmutableList.drv]: this block of duplicate of below, can we remove it by adding a return above?

**Review by yasserg — CHANGES_REQUESTED** [review]

**Comment by immutablemaps** [comment]
done all changes

**Pull request merged**